apps/sadtool/sadtool.cpp

```diff
@@ -0,0 +1,260 @@
+			printf(" \033[%dm%02X\033[0m", byte == 0 ? 2 : 1, byte);
+		}
+		printf("\n");
+	}
+
+	return EXIT_SUCCESS;
+}
+
+int ListBlocks(openblack::sad::SadFile& sad)
+{
+	auto& blocks = sad.GetBlocks();
+
+	std::printf("file: %s\n", sad.GetFilename().c_str());
+	std::printf("%u blocks\n", static_cast<uint32_t>(blocks.size()));
+	std::printf("%u sounds\n", static_cast<uint32_t>(sad.GetSounds().size()));
+	uint32_t i = 0;
+	for (auto& [name, data] : blocks)
+	{
+		std::printf("%u: name \"%s\", size %u\n", ++i, name.c_str(), static_cast<uint32_t>(data.size()));
+	}
+	std::printf("\n");
+
+	return EXIT_SUCCESS;
+}
+
+int ViewBytes(openblack::sad::SadFile& pack, const std::string& name)
+{
+	auto& block = pack.GetBlock(name);
+
+	std::printf("file: %s, block %s\n", pack.GetFilename().c_str(), name.c_str());
+
+	PrintRawBytes(block.data(), block.size() * sizeof(block[0]));
+
+	return EXIT_SUCCESS;
+}
+
+int ViewInfo(openblack::sad::SadFile& sad)
+{
+	using Lookup = openblack::sad::SadBlockHeader;
+	//	auto lookup = sad.GetInfoBlockLookup();
+	//	std::sort(lookup.begin(), lookup.end(), [](const Lookup& a, Lookup& b) { return a.blockId < b.blockId; });
+
+	std::printf("file: %s\n", sad.GetFilename().c_str());
+
+	//	for (auto& item : lookup)
+	//	{
+	//		std::printf("block: %4x, unknown: %u\n", item.blockId, item.unknown);
+	//	}
+
+	return EXIT_SUCCESS;
+}
+
+int ViewSounds(openblack::sad::SadFile& sad)
+{
+	using Lookup = openblack::sad::SadBlockHeader;
+	auto lookup = sad.GetSounds();
+	//	std::sort(lookup.begin(), lookup.end(), [](const Lookup& a, Lookup& b) { return a.blockId < b.blockId; });
+
+	std::printf("file: %s\n", sad.GetFilename().c_str());
+
+	//	for (auto& item : lookup)
+	//	{
+	//		std::printf("block: %4x, unknown: %u\n", item.blockId, item.unknown);
+	//	}
+
+	return EXIT_SUCCESS;
+}
+
+struct Arguments
+{
+	enum class Mode
+	{
+		List,
+		Bytes,
+		Info,
+		Sounds,
+		WriteSounds,
+		WriteAnimationPack,
+	};
+	std::vector<std::string> filenames;
+	Mode mode;
+	std::string block;
+	uint32_t blockId;
+	std::string outFilename;
+};
+
+bool parseOptions(int argc, char** argv, Arguments& args, int& return_code)
+{
+	cxxopts::Options options("sadtool", "Inspect and extract files from LionHead sad files.");
+
+	// clang-format off
+	options.add_options()
+		("h,help", "Display this help message.")
+		("l,list-blocks", "List block statistics.")
+		("b,view-bytes", "View raw byte content of block.", cxxopts::value<std::string>())
+		("i,info-block", "List INFO block statistics.")
+		("B,body-block", "List Body block statistics.")
+		("M,meshes-block", "List MESHES block statistics.")
+		("m,mesh", "List mesh statistics.", cxxopts::value<uint32_t>())
+		("T,texture-block", "View texture block statistics.")
+		("t,texture", "View texture statistics.", cxxopts::value<std::string>())
+		("A,animation-block", "List animation block statistics.")
+		("a,animation", "List animation statistics.", cxxopts::value<uint32_t>())
```

Comment:
These options are copy pasted from g3d tool and don't make any sense for sadtool. You should look at the options for l3dtool or anm tool